code/xcvm/cosmwasm/contracts/interpreter/src/contract.rs

```diff
@@ -0,0 +1,428 @@
+use crate::{
+	error::ContractError,
+	msg::{ExecuteMsg, InstantiateMsg, QueryMsg, XCVMInstruction, XCVMProgram},
+	state::{Config, CONFIG},
+};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+	to_binary, wasm_execute, Binary, CosmosMsg, Deps, DepsMut, Env, Event, MessageInfo,
+	QueryRequest, Response, StdError, StdResult, WasmQuery,
+};
+use cw20::{BalanceResponse, Cw20Contract, Cw20ExecuteMsg, Cw20QueryMsg};
+use num::Zero;
+use serde::Serialize;
+use std::collections::VecDeque;
+use xcvm_asset_registry::msg::{GetAssetContractResponse, QueryMsg as AssetRegistryQueryMsg};
+use xcvm_core::{Displayed, Funds, Instruction, NetworkId};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+	deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	msg: InstantiateMsg,
+) -> Result<Response, StdError> {
+	let registry_address = deps.api.addr_validate(&msg.registry_address)?;
+	let config =
+		Config { registry_address, network_id: msg.network_id, user_id: msg.user_id.clone() };
+	CONFIG.save(deps.storage, &config)?;
+
+	Ok(Response::new().add_event(
+		Event::new("xcvm.interpreter.instantiated").add_attribute(
+			"data",
+			to_binary(&(msg.network_id.0, msg.user_id))?.to_base64().as_str(),
```

Comment:
not for now, but seems it is common practice to add several attributes without join and encode. at least mars does that

Reply:
Oh, now I see. Good point. But the reason why we do this like this is only for internal use. We use this `data` attribute when handling the reply in router contract. So this won't be exposed to users or the frontend.